design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,147 @@
+# Adding and removing Ceph osds via Ceph orchestrator interface
+
+## Goals
+
+1. Give Ceph's Rook Orchestrator Plugin the ability to add and remove OSDs to/from a Rook-Ceph
+   cluster running in Kubernetes.
+2. Users of the Ceph Orchestrator CLI and/or Ceph Dashboard should be able to
+   create OSDs on nodes with as much complexity as `ceph-volume` is able to support all possible
+   user needs.
+3. Give as much room as possible for the Ceph Mgr Orchestrator Module and `ceph-volume` to implement
+   OSD management logic so that Rook's code can be simpler and long-term code maintenance on Rook
+   can be reduced.
+
+## Current status
+
+Rook expects to manage storage declaratively. Users can add disks to a Rook cluster by adding them
+to the `CephCluster` resource via the disk name itself, via 2 forms of pattern matching, or via the
+`useAllDisks` option. The Ceph Mgr Orchestrator Module differently expects to add disks to a Ceph
+cluster imperatively (not declaratively) by issuing a `ceph-volume` command with a Ceph
+[Drive Groups](https://github.com/ceph/ceph/blob/master/src/python-common/ceph/deployment/drive_group.py)
+specification. Once the `ceph-volume` command using the Drive Groups spec has been run, Ceph does
+not keep track of the Drive Groups specification used.
+
+Ceph expects the on-disk content of disks to be the declarative state of whether an OSD belongs to a
+Ceph cluster or not and what its configuration is. Rook is currently able to read this on-disk
+content via `ceph-volume` to determine if it should run a Kubernetes Deployment for a disk-based
+OSD.
+
+Rook currently does not implement OSD removal via any declarations in the `CephCluster` resource in
+order to prioritize the safety of user data. If a user wishes to remove an OSD, they must perform
+manual, imperative steps to do so.
+
+## Proposal
+
+The chief difficulty with this proposal is to harmonize interactions between Rook, which ideally
+wants to use Kubernetes manifests for a declarative intent for a Ceph cluster, and the Ceph
+Orchestrator Mgr Module, which treats OSD creation as imperative and allows the on-disk content of
+OSDs to be the declarative state.
+
+### Adding OSDs
+This design proposes to add a section to the `CephCluster` resource's `storage` configuration
+section for individual `nodes` to have disks added via a Drive Groups specification. The spec is
+proposed below with behavior described to follow.
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
+spec:
+  storage:
+    nodes: # preexisting
+      - name: < node name > # preexisting
+        driveGroups: # proposed new
+          - removeFromSpecAfterCompmletion: < default false > # proposed new
+            spec: < JSON string to pass to ceph-volume for OSD creation >
```

Comment:
The driveGroups spec it looked like it could easily be included as a yaml snippet. This would be really nice because rook users could directly add driveGroups to the cluster CR instead of needing some tool (ie. mgr module) to generate the yaml. The driveGroups really do look interesting for more than just the orchestrator module. Cluster CR admins could define driveGroups directly as they need. Internally the operator would still treat it as a yaml blob rather than attempting to deserialize it into a struct.

Reply:
Let's make sure that Drive Groups and `ceph-volume` support YAML and not just JSON.

Reply:
i've seen json converted to yaml, at least with tools